src/main/java/org/jboss/modules/ModuleLoader.java

```diff
@@ -493,13 +493,31 @@ protected final Module loadModuleLocal(String name) throws ModuleLoadException {
                 throw new ModuleLoadException("Module loader found a module with the wrong name");
             }
             final Module module;
-            if ( moduleSpec instanceof AliasModuleSpec) {
-                module = defineModule(((AliasModuleSpec) moduleSpec).asConcreteModuleSpec(), newFuture);
+            if (moduleSpec instanceof AliasModuleSpec) {
+                final AliasModuleSpec aliasModuleSpec = (AliasModuleSpec) moduleSpec;
+                Module aliasedModule;
+                while (true) {
+                    aliasedModule = loadModuleLocal(aliasModuleSpec.getAliasName());
+                    if (aliasedModule == null) {
+                        throw new ModuleLoadException("Alias module " + name + " is referencing not existing module");
+                    }
+                    final Set<String> aliases = aliasedModule.aliases;
+                    if (aliases == null) continue; // race condition detected, retry
+                    synchronized (aliases) {
+                        if (aliases == aliasedModule.aliases) {
+                            // no race-condition with removal, go ahead
+                            aliases.add(moduleSpec.getName());
+                            newFuture.setModule(module = aliasedModule);
+                            ok = true;
+                            break;
+                        }
+                    }
```

Comment:
It would be handy to have a trace statement somewhere in this if `AliasModuleSpec` branch too, perhaps just before the `break` above. Something like `log.trace("Added module %s as alias of %s from %s", moduleSpec.getName(), aliasModuleSpec.getAliasName(), this);`